core/src/main/java/org/apache/calcite/rel/AbstractRelNode.java

```diff
@@ -409,17 +407,34 @@ public RelOptTable getTable() {
    * @see #digestHash()
    */
   @API(since = "1.24", status = API.Status.EXPERIMENTAL)
-  protected boolean digestEquals(Object obj) {
+  public boolean digestEquals(Object obj) {
     if (this == obj) {
       return true;
     }
     if (this.getClass() != obj.getClass()) {
       return false;
     }
     AbstractRelNode that = (AbstractRelNode) obj;
-    return this.getTraitSet().equals(that.getTraitSet())
-        && this.getDigestItems().equals(that.getDigestItems())
+    boolean result = this.getTraitSet().equals(that.getTraitSet())
         && this.getRowType().equalsSansFieldNames(that.getRowType());
+    if (!result) {
+      return false;
+    }
+    List<Pair<String, Object>> items1 = this.getDigestItems();
+    List<Pair<String, Object>> items2 = that.getDigestItems();
+    if (items1.size() != items2.size()) {
+      return false;
+    }
+    for (int i = 0; result && i < items1.size(); i++) {
+      Pair<String, Object> attr1 = items1.get(i);
+      Pair<String, Object> attr2 = items2.get(i);
+      if (attr1.right instanceof RelNode) {
+        result = ((RelNode) attr1.right).digestEquals(attr2.right);
+      } else {
+        result = attr1.equals(attr2);
+      }
```

Comment:
we should return here, if the result is false?

Reply:
See L428. The result is checked there.